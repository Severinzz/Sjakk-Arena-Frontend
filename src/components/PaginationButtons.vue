<template>
  <div>
    <v-btn
      class="left"
      @click="jumpLeft"
      v-if="activeButton !== 1">
      &laquo;
    </v-btn>
    <v-btn
      v-for="page in visibleButtons"
      :key="page"
      :id="'button' + page"
      class="nrButton"
    @click="changePage(page)"
    :class="page === activeButton ?'active-button' : ''"
    >{{ page }}</v-btn>
    <v-btn
      class="right"
      @click="jumpRight"
    v-if="activeButton !== lastButton">
      &raquo;
    </v-btn>
      <p>Side {{ activeButton }} av {{ lastButton }}</p>
  </div>
</template>

<script>
export default {
  name: 'PaginationButtons',
  props: {
    prPage: { type: Number, required: true },
    numberOfItems: { type: Number, required: true },
    maxVisibleButtons: { type: Number, required: false, default: 5 }, // This is without the "<<" and ">>" buttons
    autoScroll: { type: Boolean, required: false, default: false }
  },
  data() {
    return {
      activeButton: 1,
      intervalId: ''
    }
  },
  methods: {
    changePage(page) {
      this.activeButton = page
      this.$emit('pageChanged', page)
    },
    // Returns if there is any buttons to the left outside of the visible ones
    hasButtonsLeft() {
      return this.minButtonNr() >= 1
    },
    // Checks if there is any buttons to the right outside of the visible ones
    hasButtonsRight() {
      return this.maxButtonNr() <= this.lastButton
    },
    // Returns the index of the lowest visible number
    minButtonNr() {
      return Math.floor(this.activeButton - (this.maxVisibleButtons / 2))
    },
    // Returns the highest number that should be visible.
    maxButtonNr() {
      return Math.floor(this.activeButton + (this.maxVisibleButtons / 2))
    },
    /*
     Jumps to the first button outside of the visible left range, if button 1 is visible that wil be jumped to.
     */
    jumpLeft() {
      if (this.minButtonNr() <= 0) {
        this.changePage(1)
      } else {
        this.changePage(this.minButtonNr())
      }
    },
    /*
    Jumps to the first button outside of the visible right range,
    if button last button is visible that wil be jumped to.
     */
    jumpRight() {
      if (this.maxButtonNr() >= this.lastButton) {
        this.changePage(this.lastButton)
      } else {
        this.changePage(this.maxButtonNr() + 1)
      }
    }
  },
  computed: {
    visibleButtons () {
      let button = 1
      let buttonsArr = []
      // Creates an array of all the buttons.
      while (buttonsArr.length < this.lastButton) {
        buttonsArr.push(button)
        button++
      }
      switch (true) {
        // If there are no buttons to the left of the visible range
        case !this.hasButtonsLeft():
          return buttonsArr.slice(0, this.maxVisibleButtons)
        // No buttons to the right of the visible range
        case !this.hasButtonsRight():
          return buttonsArr.slice(this.lastButton - this.maxVisibleButtons, this.lastButton)
        default:
          // There are buttons to both right and left of the visible range.
          return buttonsArr.slice(this.minButtonNr(), this.maxButtonNr())
      }
    },
    lastButton () {
      return Math.ceil(this.numberOfItems / this.prPage)
    }
  },
  watch: {
    autoScroll: function(autoScroll) {
      const VM = this
      if (autoScroll) {
        this.intervalId = setInterval(function () {
          VM.activeButton === VM.lastButton ? VM.activeButton = 1 : VM.activeButton = VM.activeButton + 1
          VM.changePage(VM.activeButton)
        }, 30000)
      } else {
        clearInterval(this.intervalId)
      }
    }
  }
}
</script>

<style scoped>
  p{
    margin-top: 1em;
  }
  .active-button{
    background-color: rgb(200, 200, 200) !important;
  }
</style>
