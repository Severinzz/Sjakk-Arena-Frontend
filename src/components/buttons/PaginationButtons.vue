<template>
  <div>
    <!-- [<<] button -->
    <v-btn
      class="left"
      @click="jumpLeft"
      v-if="showJumpLeft"
    >
      &laquo;
    </v-btn>
    <!-- The list of number buttons -->
    <v-btn
      v-for="page in visibleButtons"
      :key="page"
      :id="'button' + page"
      class="nrButton"
    @click="changePage(page)"
    :class="page === activeButton ?'active-button' : ''"
    >
      {{ page }}
    </v-btn>
    <!-- [>>] button -->
    <v-btn
      class="right"
      @click="jumpRight"
    v-if="showJumpRight"
    >
      &raquo;
    </v-btn>
    <!-- Current page -->
      <p>Side {{ activeButton }} av {{ lastButton }}</p>
  </div>
</template>

<script>
export default {
  name: 'PaginationButtons',
  props: {
    prPage: {
      type: Number,
      default: 5
    },
    numberOfItems: {
      type: Number,
      required: true
    },
    maxVisibleButtons: { // This is without the "<<" and ">>" buttons
      type: Number,
      default: 5
    },
    autoScroll: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      activeButton: 1,
      intervalId: '',
      showJumpLeft: false,
      showJumpRight: false
    }
  },
  methods: {

    /**
     * Change active button and emit update:page event.
     * @param page The page to change to
     */
    changePage(page) {
      this.activeButton = page
      this.$emit('update:page', page)
    },

    /**
     * Returns if there is any buttons to the left outside of the visible ones
     */
    hasButtonsLeft() {
      return this.minButtonNr() >= 1
    },

    /**
     * Checks if there is any buttons to the right outside of the visible ones
     */
    hasButtonsRight() {
      return this.maxButtonNr() < this.lastButton
    },

    /**
     * Returns the index of the lowest visible number
     */
    minButtonNr() {
      return Math.floor(this.activeButton - (this.maxVisibleButtons / 2))
    },

    /**
     * Returns the highest number that should be visible.
     */
    maxButtonNr() {
      return Math.floor(this.activeButton + (this.maxVisibleButtons / 2))
    },

    /**
     * Jumps to the first button outside of the visible left range, if button 1 is visible that wil be jumped to.
     */
    jumpLeft() {
      if (this.minButtonNr() <= 0) {
        this.changePage(1)
      } else {
        this.changePage(this.minButtonNr())
      }
    },

    /**
     * Jumps to the first button outside of the visible right range,
     * if button last button is visible that wil be jumped to.
     */
    jumpRight() {
      if (this.maxButtonNr() >= this.lastButton) {
        this.changePage(this.lastButton)
      } else {
        this.changePage(this.maxButtonNr() + 1)
      }
    },

    /**
     * Hides or displays the [<<] button.
     * @param show Boolean value to deside if [<<] should be shown
     */
    alterShowJumpLeft(show) {
      this.showJumpLeft = show
    },

    /**
     * Hides or displays the [>>] button.
     * @param show Boolean value to deside if [>>] should be shown
     */
    alterShowJumpRight(show) {
      this.showJumpRight = show
    }
  },
  computed: {

    /**
     * Returns an array consisting of the number of the buttons that should be visible.
     * @returns {[]|*[]} Array consisting of the number of the buttons that should be visible.
     */
    visibleButtons () {
      let button = 1
      let buttonsArr = []
      // Creates an array of all the buttons.
      while (buttonsArr.length < this.lastButton) {
        buttonsArr.push(button)
        button++
      }
      switch (true) {
        case this.maxVisibleButtons >= this.lastButton:
          this.alterShowJumpLeft(false)
          this.alterShowJumpRight(false)
          return buttonsArr

        // If there are no buttons to the left of the visible range
        case !this.hasButtonsLeft():
          this.alterShowJumpLeft(false)
          this.alterShowJumpRight(true)
          return buttonsArr.slice(0, this.maxVisibleButtons)

        // No buttons to the right of the visible range
        case !this.hasButtonsRight():
          this.alterShowJumpLeft(true)
          this.alterShowJumpRight(false)
          return buttonsArr.slice(this.lastButton - this.maxVisibleButtons, this.lastButton)

        default:
          this.alterShowJumpLeft(true)
          this.alterShowJumpRight(true)
          // There are buttons to both right and left of the visible range.
          return buttonsArr.slice(this.minButtonNr(), this.maxButtonNr())
      }
    },

    /**
     * Returns the last button
     * @returns {number} The last button
     */
    lastButton () {
      let lastButton = Math.ceil(this.numberOfItems / this.prPage)
      return lastButton > 0 ? lastButton : 1
    }
  },
  watch: {

    /**
     * Go the the next button every 30 sec. When the last button is reached, start from the beginning.
     * @param autoScroll Boolean value to tell if it should auto scroll or not.
     */
    autoScroll: function(autoScroll) {
      const VM = this
      if (autoScroll) {
        this.intervalId = setInterval(function () {
          VM.activeButton === VM.lastButton ? VM.activeButton = 1 : VM.activeButton = VM.activeButton + 1
          VM.changePage(VM.activeButton)
        }, 30000)
      } else {
        clearInterval(this.intervalId)
      }
    }
  },
  destroyed() {
    clearInterval(this.intervalId)
  }
}
</script>

<style scoped>
  p{
    margin-top: 1em;
  }
  .active-button{
    background-color: rgb(200, 200, 200) !important;
  }
</style>
