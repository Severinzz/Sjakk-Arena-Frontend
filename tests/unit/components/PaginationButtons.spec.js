import { createLocalVue, mount } from '@vue/test-utils'
import Vue from 'vue'
import Vuetify from 'vuetify'

import PaginationButtons from '../../../src/components/PaginationButtons'

Vue.use(Vuetify)

const localVue = createLocalVue()

const factory = (propsData) => {
  return mount(PaginationButtons, {
    localVue,
    propsData: {
      ...propsData
    }
  })
}

describe('PaginationButtons', () => {

  it('should emit event with the number of the clicked button', async() => {
    const wrapper = factory({
      numberOfItems: 20,
      prPage: 3 })
    wrapper.find('#button5').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('pageChanged')[0]).toEqual([5])
  })
  it('should match snapshot', () => {
    const wrapper = factory({
      numberOfItems: 20,
      prPage: 3 })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('Jump left button should be visible after page change to right', async() => {
    const wrapper = factory({
      numberOfItems: 20,
      prPage: 3 })
    wrapper.find('.right').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.find('.left').isVisible()).toBe(true)
  })
  it('should render the correct active button', () => {
    const wrapper = factory({
      numberOfItems: 20,
      prPage: 3 })
    wrapper.setData({ activeButton: 6 })
    expect(wrapper.find('.active-button').text()).toBe('6')
  })
  it('should render the correct amount of buttons', () => {
    // NOTE: maxVisibleButtons does NOT count "<<" and ">>" buttons!
    const visibleButtons = 3
    const wrapper = factory({
      numberOfItems: 60,
      prPage: 3,
      maxVisibleButtons: visibleButtons
    })
    wrapper.setData({ activeButton: 10 })
    expect(wrapper.findAll('.nrButton').length).toBe(visibleButtons)
  })
})
