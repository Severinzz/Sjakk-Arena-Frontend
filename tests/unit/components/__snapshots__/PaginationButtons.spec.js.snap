// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationButtons should match snapshot 1`] = `
<div>
  <!----> <button type="button" class="nrButton v-btn v-btn--contained theme--light v-size--default active-button" id="button1"><span class="v-btn__content">1</span></button><button type="button" class="nrButton v-btn v-btn--contained theme--light v-size--default" id="button2"><span class="v-btn__content">2</span></button><button type="button" class="nrButton v-btn v-btn--contained theme--light v-size--default" id="button3"><span class="v-btn__content">3</span></button><button type="button" class="nrButton v-btn v-btn--contained theme--light v-size--default" id="button4"><span class="v-btn__content">4</span></button><button type="button" class="nrButton v-btn v-btn--contained theme--light v-size--default" id="button5"><span class="v-btn__content">5</span></button> <button type="button" class="right v-btn v-btn--contained theme--light v-size--default"><span class="v-btn__content">
    Â»
  </span></button>
  <p>Side 1 av 7</p>
</div>
`;
